src/generation/viterbi/observation_function.py

```diff
@@ -0,0 +1,71 @@
+from typing import Callable, List, FrozenSet
+import numpy as np
+from numpy.typing import NDArray
+
+from common.note_collection import NoteCollection
+from common.structures.pitch import Pitch
+
+from generation.viterbi.viterbi_index import ViterbiIndex
+
+
+class ObservationFunction:
+    """
+    A function format of the observation matrix for the Viterbi chord progression generator.
+    This function represents the probability of observing a particular pitch given a chord.
+
+    :param algorithm: The algorithm to use for calculating the probability. By default, this
+        is the frequency counter algorithm.
+    """
+
+    def __init__(
+        self,
+        algorithm: Callable[[int, List[FrozenSet[Pitch]], int], float] | None = None,
```

Comment:
It'd be nice to make a type definition for `Callable[[int, List[FrozenSet[Pitch]], int], float]`.